PhysicsTools/IsolationAlgos/plugins/CITKPFIsolationSumProducerForPUPPI.cc

```diff
@@ -0,0 +1,182 @@
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/MuonReco/interface/MuonFwd.h"
+#include "DataFormats/RecoCandidate/interface/IsoDepositFwd.h"
+#include "DataFormats/RecoCandidate/interface/IsoDeposit.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+#include "PhysicsTools/IsolationAlgos/interface/EventDependentAbsVeto.h"
+#include "DataFormats/Candidate/interface/CandidateFwd.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "PhysicsTools/IsolationAlgos/interface/CITKIsolationConeDefinitionBase.h"
+#include "DataFormats/Common/interface/OwnVector.h"
+
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+
+#include <string>
+#include <unordered_map>
+
+//module to compute isolation sum weighted with PUPPI weights
+namespace citk {
+  class PFIsolationSumProducerForPUPPI : public edm::stream::EDProducer<> {
+    
+  public:  
+    PFIsolationSumProducerForPUPPI(const edm::ParameterSet&);
+    
+    virtual ~PFIsolationSumProducerForPUPPI() {}
+    
+    virtual void beginLuminosityBlock(const edm::LuminosityBlock&,
+			      const edm::EventSetup&) override final;
+
+    virtual void produce(edm::Event&, const edm::EventSetup&) override final;
+    
+  private:  
+    // datamembers
+    static constexpr unsigned kNPFTypes = 8;
+    typedef std::unordered_map<std::string,int> TypeMap;
+    typedef std::vector<std::unique_ptr<IsolationConeDefinitionBase> > IsoTypes;
+    typedef edm::View<reco::Candidate> CandView;
+    const TypeMap _typeMap;
+    edm::EDGetTokenT<CandView> _to_isolate, _isolate_with;
+    edm::EDGetTokenT<edm::ValueMap<float> > puppiValueMapToken_;//for puppiValueMap
+    edm::Handle<edm::ValueMap<float>> puppiValueMap;//puppiValueMap
+    // indexed by pf candidate type
+    std::array<IsoTypes,kNPFTypes> _isolation_types; 
+    std::array<std::vector<std::string>,kNPFTypes> _product_names;
+    bool useValueMapForPUPPI = true;
+  };
+}
+
+typedef citk::PFIsolationSumProducerForPUPPI CITKPFIsolationSumProducerForPUPPI;
+
+DEFINE_FWK_MODULE(CITKPFIsolationSumProducerForPUPPI);
+
+namespace citk {
+  PFIsolationSumProducerForPUPPI::PFIsolationSumProducerForPUPPI(const edm::ParameterSet& c) :
+    _typeMap( { {"h+",1},
+	        {"h0",5},
+		{"gamma",4},
+		{"electron",2},
+		{"muon",3},
+		{"HFh",6},
+		{"HFgamma",7} } ){
+    _to_isolate = 
+      consumes<CandView>(c.getParameter<edm::InputTag>("srcToIsolate"));
+    _isolate_with = 
+      consumes<CandView>(c.getParameter<edm::InputTag>("srcForIsolationCone"));
+      if (c.exists("puppiValueMap")) {
```

Comment:
use of  ParameterSet exists is discouraged.
I can suggest using an empty InputTag ("") instead. You can use the same logic here then.